streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+        // we ensure that there are pending transaction and verify that data is processed correctly.
+        //
+        // We either close clients cleanly (`injectError = false`) or let them crash (`injectError = true`) during
+        // the upgrade. For both cases, EOS should not be violated.
+        //
+        // The commit interval is set to MAX_VALUE and the used `Processor` request commits manually so we have full
+        // control when a commit actually happens. We use an input topic with 2 partitions and each task will request
+        // a commit after processing 10 records.
+        //
+        // 1.  start both clients and wait until rebalance stabilizes
+        // 2.  write 10 records per input topic partition and verify that the result was committed
+        // 3.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - both pending transactions are based on task producers
+        // 4.  stop/crash the first client, wait until rebalance stabilizes:
+        //      - stop case:
+        //        * verify that the stopped client did commit its pending transaction during shutdown
+        //        * the second client will still have an pending transaction
+        //      - crash case:
+        //        * the second client will have two pending transactions
+        // 5.  restart the first client with eos-beta enabled and wait until rebalance stabilizes
+        //       - crash only: the rebalance should result in a commit
+        // 6.  write 5 record per input topic partition and verify that the result was committed
+        // 7.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - one transaction is base on a task producer; the other transaction is based on a thread producer
+        // 8.  stop/crash the second client and wait until rebalance stabilizes:
+        //      - stop only: verify that the stopped client did commit its pending transaction during shutdown
+        //      - the first client will still have an pending transaction
+        // 9.  restart the second client with eos-beta enabled and wait until rebalance stabilizes
+        // 10. write 5 record per input topic partition and verify that the result was committed
+
+        try (final KafkaStreams streams1Alpha = getKafkaStreams("appDir1", StreamsConfig.EXACTLY_ONCE);
+            final KafkaStreams streams1Beta = getKafkaStreams("appDir1", StreamsConfig.EXACTLY_ONCE_BETA);
+            final KafkaStreams streams2Alpha = getKafkaStreams("appDir2", StreamsConfig.EXACTLY_ONCE);
+            final KafkaStreams streams2Beta = getKafkaStreams("appDir2", StreamsConfig.EXACTLY_ONCE_BETA)) {
+
+            final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> stateTransitions1 = new LinkedList<>();
```

Comment:
Should we make this a concurrent linked list since concurrent threads may be adding at the same time?

Reply:
There should not be any concurrent threads, because `stream1Alpha` and `streams1Beta` don't run at the same time but strictly one after each other.

Reply:
Ah, you're right.